app/code/Magento/Downloadable/Controller/Download/Sample.php

```diff
@@ -46,7 +46,9 @@ public function execute()
         $sampleId = $this->getRequest()->getParam('sample_id', 0);
         /** @var \Magento\Downloadable\Model\Sample $sample */
         $sample = $this->_objectManager->create(\Magento\Downloadable\Model\Sample::class)->load($sampleId);
-        if ($sample->getId() && $this->salabilityChecker->isSalable($sample->getProductId())) {
+        /** @var \Magento\Catalog\Model\Product $product */
+        $product = $this->productRepository->getById($sample->getProductId());
+        if ($sample->getId() && $product->isInStock()) {
```

Comment:
Would be great to not depend on non-interface method. I think it will be enough to check `(int)$product->getStatus() === \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED`.


Questions: What behavior will we have if:
- product isn't associated to this website?
- product is "disabled"?
- product visibility set to "Not visible individually"?

AFAIK In all these 3 cases product will not be available on frontend -> downaload sample should not be available either.

Reply:
I have changed the functionality, Please check accordingly.